Add support for global fragments

mpourismaiel opened the pull request:
<!--
Thanks for sending a pull request!  
If this is your first time, read our [contributing guidelines](/CONTRIBUTING.md).
-->
**What this PR does / why we need it**:
Users now can add more used fragments into global and the theme will automatically use that global fragment if no local fragment is overriding it.

**Which issue this PR fixes** *(optional - uncomment and add issue)*:
fixes #92

stp-ip commented on layouts/_default/single.html:
```diff
@@ -9,51 +9,32 @@
+    fragment is shown. The same is true if there are no global footer fragments
+    and one local one. But if there is one global footer fragment and one local
+    footer fragment, the local one is rendered.
+    */}}
```
Thanks for documenting. Would love some thoughts from @lauragift21.
I have the design in my head so it seems fine. I would probably clarify the difference between "fragment" and the custom fragment name aka filename. As the filename would be the factor in overwriting not the "fragment" type/name.

lauragift21 replied:
Personally, I think going with the global fragment method is way better. It simplifies the layout and is a lot more reusable in terms of adding another "fragment", that brings me to ask do we still need to add the split layout using this method or everything is just going to be within a single layout?

lauragift21 replied:
Alright. this explains it better.

stp-ip commented on layouts/_default/single.html:
```diff
@@ -9,51 +9,32 @@
-      {{ partial (print "fragments/" .Params.fragment_name ".html") (dict "self" .) }}
+  {{ range sort ($.Scratch.Get "fragments") "Params.weight" }}
+    {{ if and (not .Params.disabled) (isset .Params "fragment") (not (eq .Params.fragment "hero")) }}
+      {{ partial (print "fragments/" .Params.fragment ".html") (dict "menus" $.Site.Menus "Site" $.Site "resources" $.Resources "self" .) }}
```
I would have assumed we used $.Scratch in the fragments so that not providing "$" or "tmp_full" would break hero, table and item.

stp-ip replied:
Considering this we might actually just move to "$" and "." in this PR. Thoughts?

mpourismaiel replied:
Let's continue this one in the other discussion https://github.com/okkur/syna/pull/108#discussion_r187721597

stp-ip commented on layouts/_default/single.html:
```diff
@@ -9,51 +9,37 @@
 
-    {{ if and (not .Params.disabled) (in .Params.fragment "custom") }}
-      {{ partial (print "fragments/" .Params.fragment_name ".html") (dict "self" .) }}
+  {{ range sort ($.Scratch.Get "fragments" | default slice) "Params.weight" }}
```
What's the reasoning behind default slice?

mpourismaiel replied:
Sometimes there are simply no fragments. For example nesting page has no local fragments and if you remove global fragments the loop would crash since `$.Scratch.Get "fragments"` would be `nil`. `default slice` seemed cleaner than an if block.

stp-ip replied:
Thanks. Makes sense.

stp-ip commented on layouts/_default/single.html:
```diff
@@ -9,51 +9,37 @@
-      {{ partial (print "fragments/" .Params.fragment_name ".html") (dict "self" .) }}
+  {{ range sort ($.Scratch.Get "fragments" | default slice) "Params.weight" }}
+    {{ if and (not .Params.disabled) (isset .Params "fragment") (not (eq .Params.fragment "hero")) }}
+      {{ partial (print "fragments/" .Params.fragment ".html") (dict "menus" $.Site.Menus "Site" $.Site "resources" $.Resources "self" .) }}
```
Fascinating we are using $.Scratch in hero (PR), table and item, but seem not to run into issues. Have to check that out.

mpourismaiel replied:
$.Scratch is scoped and can be used inside partials with their own scopes so not passing the $ for top level in pages to partials doesn't make $.Scratch unusable since $ in partials refers to their own top level scope.

stp-ip commented:
Let's fix the broken build and then add the same logic to the split layout. Then we should be able to merge.
I'll be doing some testing tomorrow until we have the above things done.

mpourismaiel commented:
I fixed the conflict and the issue with build (thanks @stp-ip). Also added the same code to `split` layout, although I changed the code to match the previous way split was rendered (weirdly, hero is rendered after content in there which would be fixed if we decide to make content partials fragments as well). I think everything is in order.

stp-ip commented:
Great. The "broken" split layout will be fixed with having content a fragment. For now it doesn't break the example page and it won't be released without content being a fragment.

/LGTM
